File: src/main/java/org/moinex/ui/dialog/investment/EditBondPurchaseController.java
/*
 * Filename: EditBondPurchaseController.java
 * Created on: January  3, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog.investment;

import jakarta.persistence.EntityNotFoundException;
import java.math.BigDecimal;
import java.time.LocalDate;
import javafx.fxml.FXML;
import javafx.stage.Stage;
import lombok.NoArgsConstructor;
import org.moinex.model.Category;
import org.moinex.model.enums.TransactionStatus;
import org.moinex.model.enums.TransactionType;
import org.moinex.model.investment.BondOperation;
import org.moinex.model.wallettransaction.Wallet;
import org.moinex.service.BondService;
import org.moinex.service.CategoryService;
import org.moinex.service.I18nService;
import org.moinex.service.WalletService;
import org.moinex.service.WalletTransactionService;
import org.moinex.util.Constants;
import org.moinex.util.WindowUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;

@Controller
@NoArgsConstructor
public final class EditBondPurchaseController extends BaseBondTransactionManagement {
    private BondOperation operation = null;

    @Autowired
    public EditBondPurchaseController(
            WalletService walletService,
            WalletTransactionService walletTransactionService,
            CategoryService categoryService,
            BondService bondService,
            I18nService i18nService) {
        super(walletService, walletTransactionService, categoryService, bondService, i18nService);
        this.i18nService = i18nService;
        transactionType = TransactionType.EXPENSE;
    }

    public void setOperation(BondOperation op) {
        this.operation = op;
        this.bond = op.getBond();

        String symbol = bond.getSymbol();
        bondNameLabel.setText(
                bond.getName() + (symbol != null && !symbol.isBlank() ? " (" + symbol + ")" : ""));

        setWalletComboBox(operation.getWalletTransaction().getWallet());

        descriptionField.setText(operation.getWalletTransaction().getDescription());
        unitPriceField.setText(operation.getUnitPrice().toString());
        quantityField.setText(operation.getQuantity().toString());

        if (operation.getFees() != null) {
            feesField.setText(operation.getFees().toString());
        }

        if (operation.getTaxes() != null) {
            taxesField.setText(operation.getTaxes().toString());
        }

        statusComboBox.setValue(operation.getWalletTransaction().getStatus());
        categoryComboBox.setValue(operation.getWalletTransaction().getCategory());
        transactionDatePicker.setValue(operation.getWalletTransaction().getDate().toLocalDate());

        updateTotalPrice();
    }

    @FXML
    @Override
    protected void handleSave() {
        Wallet wallet = walletComboBox.getValue();
        String description = descriptionField.getText();
        TransactionStatus status = statusComboBox.getValue();
        Category category = categoryComboBox.getValue();
        String unitPriceStr = unitPriceField.getText();
        String quantityStr = quantityField.getText();
        String feesStr = feesField.getText();
        String taxesStr = taxesField.getText();
        LocalDate buyDate = transactionDatePicker.getValue();

        if (wallet == null
                || description == null
                || description.isBlank()
                || status == null
                || category == null
                || unitPriceStr == null
                || unitPriceStr.isBlank()
                || quantityStr == null
                || quantityStr.isBlank()
                || buyDate == null) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_EMPTY_FIELDS_TITLE),
                    i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_EMPTY_FIELDS_MESSAGE));

            return;
        }

        try {
            BigDecimal unitPrice = new BigDecimal(unitPriceStr);
            BigDecimal quantity = new BigDecimal(quantityStr);
            BigDecimal fees =
                    (feesStr != null && !feesStr.isBlank()) ? new BigDecimal(feesStr) : null;
            BigDecimal taxes =
                    (taxesStr != null && !taxesStr.isBlank()) ? new BigDecimal(taxesStr) : null;

            boolean feesEqual =
                    (operation.getFees() == null && fees == null)
                            || (operation.getFees() != null
                                    && fees != null
                                    && operation.getFees().compareTo(fees) == 0);
            boolean taxesEqual =
                    (operation.getTaxes() == null && taxes == null)
                            || (operation.getTaxes() != null
                                    && taxes != null
                                    && operation.getTaxes().compareTo(taxes) == 0);

            if (operation.getWalletTransaction().getWallet().getId().equals(wallet.getId())
                    && operation.getWalletTransaction().getDescription().equals(description)
                    && operation.getWalletTransaction().getStatus().equals(status)
                    && operation
                            .getWalletTransaction()
                            .getCategory()
                            .getId()
                            .equals(category.getId())
                    && operation.getUnitPrice().compareTo(unitPrice) == 0
                    && operation.getQuantity().compareTo(quantity) == 0
                    && feesEqual
                    && taxesEqual
                    && operation.getWalletTransaction().getDate().toLocalDate().equals(buyDate)) {
                WindowUtils.showInformationDialog(
                        i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_NO_CHANGES_TITLE),
                        i18nService.tr(
                                Constants.TranslationKeys.BOND_DIALOG_NO_CHANGES_PURCHASE_MESSAGE));
            } else {
                bondService.updateOperation(
                        operation.getId(),
                        wallet.getId(),
                        quantity,
                        unitPrice,
                        buyDate,
                        fees,
                        taxes,
                        null,
                        category,
                        description,
                        status);

                WindowUtils.showSuccessDialog(
                        i18nService.tr(
                                Constants.TranslationKeys.BOND_DIALOG_PURCHASE_UPDATED_TITLE),
                        i18nService.tr(
                                Constants.TranslationKeys.BOND_DIALOG_PURCHASE_UPDATED_MESSAGE));
            }

            Stage stage = (Stage) bondNameLabel.getScene().getWindow();
            stage.close();
        } catch (NumberFormatException e) {
            WindowUtils.showErrorDialog(
                    i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_INVALID_NUMBER_TITLE),
                    i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_INVALID_NUMBER_MESSAGE));
        } catch (EntityNotFoundException | IllegalArgumentException e) {
            WindowUtils.showErrorDialog(
                    i18nService.tr(
                            Constants.TranslationKeys.BOND_DIALOG_ERROR_UPDATING_PURCHASE_TITLE),
                    e.getMessage());
        }
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/investment/update/EditBondPurchaseController.kt
/*
 * Filename: EditBondPurchaseController.kt (original filename: EditBondPurchaseController.java)
 * Created on: January  3, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 20/03/2026
 */

package org.moinex.ui.dialog.investment.update

import jakarta.persistence.EntityNotFoundException
import javafx.fxml.FXML
import javafx.stage.Stage
import org.moinex.common.constant.TranslationKeys
import org.moinex.common.extension.isEqual
import org.moinex.common.util.WindowUtils
import org.moinex.model.dto.WalletTransactionContextDTO
import org.moinex.model.enums.WalletTransactionType
import org.moinex.model.investment.BondOperation
import org.moinex.service.CategoryService
import org.moinex.service.PreferencesService
import org.moinex.service.investment.BondService
import org.moinex.service.wallet.WalletService
import org.moinex.ui.dialog.investment.base.BaseBondTransactionManagement
import org.springframework.stereotype.Controller
import java.math.BigDecimal

@Controller
class EditBondPurchaseController(
    walletService: WalletService,
    categoryService: CategoryService,
    bondService: BondService,
    preferencesService: PreferencesService,
) : BaseBondTransactionManagement(
        walletService,
        categoryService,
        bondService,
        preferencesService,
    ) {
    private lateinit var operation: BondOperation

    init {
        walletTransactionType = WalletTransactionType.EXPENSE
    }

    fun setOperation(op: BondOperation) {
        operation = op
        bond = op.bond

        val symbol = bond.symbol
        bondNameLabel.text = bond.name + if (!symbol.isNullOrBlank()) " ($symbol)" else ""

        setWalletComboBox(operation.walletTransaction!!.wallet)

        descriptionField.text = operation.walletTransaction!!.description
        unitPriceField.text = operation.unitPrice.toString()
        quantityField.text = operation.quantity.toString()
        feesField.text = operation.fees.toString()
        taxesField.text = operation.taxes.toString()

        statusComboBox.value = operation.walletTransaction!!.status
        categoryComboBox.value = operation.walletTransaction!!.category
        transactionDatePicker.value = operation.walletTransaction!!.date.toLocalDate()
        includeInAnalysisCheckBox.isSelected = operation.walletTransaction!!.includeInAnalysis

        updateTotalPrice()
    }

    @FXML
    override fun handleSave() {
        val wallet = walletComboBox.value
        val description = descriptionField.text
        val status = statusComboBox.value
        val category = categoryComboBox.value
        val unitPriceStr = unitPriceField.text
        val quantityStr = quantityField.text
        val feesStr = feesField.text
        val taxesStr = taxesField.text
        val buyDate = transactionDatePicker.value

        if (wallet == null ||
            description.isNullOrBlank() ||
            status == null ||
            category == null ||
            unitPriceStr.isNullOrBlank() ||
            quantityStr.isNullOrBlank() ||
            buyDate == null
        ) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.BOND_DIALOG_EMPTY_FIELDS_TITLE),
                preferencesService.translate(TranslationKeys.BOND_DIALOG_EMPTY_FIELDS_MESSAGE),
            )
            return
        }

        runCatching {
            val unitPrice = unitPriceStr.toBigDecimal()
            val quantity = quantityStr.toBigDecimal()
            val fees = if (!feesStr.isNullOrBlank()) feesStr.toBigDecimal() else BigDecimal.ZERO
            val taxes = if (!taxesStr.isNullOrBlank()) taxesStr.toBigDecimal() else BigDecimal.ZERO

            val feesEqual = operation.fees.isEqual(fees)
            val taxesEqual = operation.taxes.isEqual(taxes)

            if (operation.walletTransaction!!.wallet.id == wallet.id &&
                operation.walletTransaction!!.description == description &&
                operation.walletTransaction!!.status == status &&
                operation.walletTransaction!!.category.id == category.id &&
                operation.walletTransaction!!.includeInAnalysis == includeInAnalysisCheckBox.isSelected &&
                operation.unitPrice.isEqual(unitPrice) &&
                operation.quantity.isEqual(quantity) &&
                feesEqual &&
                taxesEqual &&
                operation.walletTransaction!!.date.toLocalDate() == buyDate
            ) {
                WindowUtils.showInformationDialog(
                    preferencesService.translate(TranslationKeys.BOND_DIALOG_NO_CHANGES_TITLE),
                    preferencesService.translate(TranslationKeys.BOND_DIALOG_NO_CHANGES_PURCHASE_MESSAGE),
                )
            } else {
                operation.quantity = quantity
                operation.unitPrice = unitPrice
                operation.fees = fees
                operation.taxes = taxes

                bondService.updateBondOperation(
                    operation,
                    WalletTransactionContextDTO(
                        wallet,
                        buyDate.atStartOfDay(),
                        category,
                        description,
                        status,
                        includeInAnalysisCheckBox.isSelected,
                    ),
                )
            }

            (bondNameLabel.scene.window as Stage).close()
        }.onFailure { e ->
            when (e) {
                is NumberFormatException -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(TranslationKeys.BOND_DIALOG_INVALID_NUMBER_TITLE),
                        preferencesService.translate(TranslationKeys.BOND_DIALOG_INVALID_NUMBER_MESSAGE),
                    )
                }
                is EntityNotFoundException, is IllegalArgumentException -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(TranslationKeys.BOND_DIALOG_ERROR_UPDATING_PURCHASE_TITLE),
                        e.message ?: "",
                    )
                }
                else -> throw e
            }
        }
    }
}
